<span style='line-height:1.5;font-size: 17px;font-family:"Times New Roman"'>


<b><center><font size="6">Contrôle continu 2 - Econométrie en finance</font></center></b>
    
<b><center><font size="5"> Sujet 1</font></center></b>
    
<g><h7><center> CM - P. Durand  - pierre.durand@u-pec.fr </center></h7></g>

<g><h7><center> TD - R.Birem  - abderraouf.birem@u-pec.fr </center></h7></g>


<b><center><font size="5">M1 MBFA - 2023/2024</font></center></b> 



</span>

<span style='line-height:1.5;font-size: 17px;font-family:"Times New Roman"'>

<b><center><font size="5">Veuillez insérer votre Nom, Prénom et Numéro d`étudiant</font></center></b> 

</span>

In [1]:
#Nom : KACHAOU
#Prénom : Issa

In [2]:
Num_Etudiant=21905943

<span style='line-height:1.5;font-size: 17px;font-family:"Times New Roman"'>



Epreuves : 
    
- Le sujet dépendra du numéro que vous insérerez ci-dessous.
- Les copies doivent être envoyées par courrier électronique cinq minutes avant la fin du cours à l'adresse abderraouf.birem@u-pec.fr. Chaque minute de retard entraînera une pénalité d'un point.
   

    
    
<br>
    
    


</span>

<span style='line-height:1.5;font-size: 17px;font-family:"Times New Roman"'>

<b><center><font size="5">l'exécution des cellules suivantes est obligatoire pour la suite</font></center></b> 

</span>

In [3]:
Vars = ['Accor', 'Airbus', 'AirLiquide', 'URWestfield', 'Atos', 'BNP',
       'Carrefour', 'Capgemini', 'AXA', 'Vinci', 'Dassault', 'Loreal', 'Engie'
        , 'SocieteGenerale', 'Thales', 'Kering']

In [6]:
import random
def convert_to_range(number):
    if len(str(number)) != 8:
        return 
    random.seed(number)
    result = random.randint(0, 15)
    return result
code = convert_to_range(Num_Etudiant)

<span style='line-height:1.5;font-size: 17px;font-family:"Times New Roman"'>

<b><center><font size="5">Importation des packages</font></center></b> 

</span>

In [5]:
import pandas as pd

<span style='line-height:1.5;font-size: 17px;font-family:"Times New Roman"'>

<b><center><font size="5">Importation des données</font></center></b> 

</span>

In [7]:
chemin = "/Users/anissaouaghlani/Desktop/M1 MBFA II/Econométrie/"
df_temp = pd.read_csv(chemin + "base_cc2.csv" , sep=';')

<span style='line-height:1.5;font-size: 17px;font-family:"Times New Roman"'>

<b><center><font size="5">Générez la base finale pour l'étude en exécutant la prochaine cellule</font></center></b> 

</span>

In [9]:
df = df_temp[['Date', 'CAC40', 'TxChange', Vars[code]]]
df

Date        CAC40  TxChange  SocieteGenerale
0    2010-01-01  3739.459961  1.387694        20.472494
1    2010-02-01  3708.800049  1.363791        19.596941
2    2010-03-01  3974.010010  1.350804        22.587410
3    2010-04-01  3816.989990  1.329805        19.596941
4    2010-05-01  3507.560059  1.228788        17.001795
..          ...          ...       ...              ...
161  2023-06-01  7400.060059  1.086803        23.799999
162  2023-07-01  7497.779785  1.102426        24.715000
163  2023-08-01  7316.700195  1.093255        26.215000
164  2023-09-01  7135.060059  1.058761        23.045000
165  2023-10-01  6885.649902  1.061504        21.145000

[166 rows x 4 columns]

<span style='line-height:1.5;font-size: 26px;font-family:"Times New Roman"'>
<hr style=" border: 2.5px solid black;">
En utilisant la base de données "df", effectuez une étude économétrique complète visant à évaluer la rentabilité d'une entreprise française en fonction des fluctuations du taux de change Euro/dollar et de l'évolution de l'indice CAC 40. Mettez en œuvre l'ensemble de vos connaissances acquises durant le semestre. 
    
                                                    Bonne chance

<hr style=" border: 2.5px solid black;">   
    
<br>
    
</span>

## Statistiques decriptves 

In [10]:
df.describe()

CAC40    TxChange  SocieteGenerale
count   166.000000  166.000000       166.000000
mean   4928.521876    1.199356        19.646374
std    1132.559930    0.114769         5.148377
min    2981.959961    0.982956         8.187415
25%    4031.934937    1.111775        16.973186
50%    4796.820068    1.170618        20.427595
75%    5520.099976    1.304295        23.069016
max    7497.779785    1.482096        29.882158

- Nous pouvons voir que la moyenne du CAC40 est de 4928,52, ce qui indique le cours moyen de l'indice boursier. La volatilité, reflétée par l'écart type, est de 1132,55, suggérant  une certaine variabilité des 40 actions les plus importantes. 
- Concernant le taux de change, la moyenne est de 1,19 avec un ecart type de 0,11 ce qui nous informe sur une très faible volatilité
- La variable Société Générale, qui est la variable expliquée, a une rentabilité en moyenne de 19,64 avec un écart type de 5,14. Cette volatilité s'explique par la forte volatilité du CAC 40. 
- Les valeurs minimales (min) et maximales (max) nous donne des informations sur les bornes pour chaque variable. Pour le CAC 40 [7497,7; 2981] alors que pour le taux de change [1,48 ; 0,98]

In [11]:
df.corr()

/var/folders/sg/frcwz3bs4nqf585cn0brfmjw0000gn/T/ipykernel_1012/1134722465.py:1: FutureWarning: The default value of numeric_only in DataFrame.corr is deprecated. In a future version, it will default to False. Select only valid columns or specify the value of numeric_only to silence this warning.
  df.corr()


CAC40  TxChange  SocieteGenerale
CAC40            1.000000 -0.671789         0.503349
TxChange        -0.671789  1.000000        -0.246643
SocieteGenerale  0.503349 -0.246643         1.000000

- Le coefficient de corrélation entre le CAC 40 et le taux de change est de -0,671. Cette corrélation négative suggère une relation inverse modérée. Lorsque le cours de l'indice boursier du CAC40 augmente, le taux de change a tendance à diminuer, et vice versa.
- Le coefficient entre le CAC40 et l'entreprise financière Societe Generale est de 0,503. Cela suggère une corrélation positive modérée entre ces deux variables, indiquant que, en général, ils ont tendance à évoluer dans la même direction.
- Pour finir, le coefficient de corrélation entre le taux de change et societe generale est de -0,246. Encore une fois nous avons une correlation négative entre les 2 variables ce qui nous indiques que lorsque ce que le taux de change augmente, la rentabilité de societe generale baisse. 

## Graphiques

In [13]:
from statsmodels.graphics.tsaplots import plot_acf
from statsmodels.graphics.tsaplots import plot_pacf

import matplotlib.pyplot as plt 

## Analyse de le stationnarité 

Pourquoi faut-il que les séries soient stationnaires ?
- Pas de stationnarité ⇒ propriétés asymptotiques non valables (estimateurs à forte variance)
- Pas de stationnarité ⇒ possibilité de régressions fallacieuses (cas extrême des : corrélations fallacieuses

In [15]:
from statsmodels.tsa.stattools import adfuller

<span style='line-height:1.5;font-size: 17px;font-family:"Times New Roman"'>



- Valeurs critiques de la constante et de la tendance dans le test de Dickey-Fuller

<table>
    <col width="50">
    <col width="70">
    <col width="70">
    <col width="70">
    <col width="70">
    <col width="70">
    <col width="70">
    <col width="70">
    <col width="70">
    <col width="70">
    <tr>
        <td colspan="1" style="text-align: center;border-right: 2px solid black"></td>
        <td colspan="3" style="text-align: center;border-right: 2px solid black"><i>Modèle [2]</i></td>
        <td colspan="6" style="text-align: center;border-right: 2px solid black"><i>Modèle [1]</i></td>
    </tr>
    <tr>
        <td colspan="1" style="text-align: center;border-right: 2px solid black"></td>
        <td colspan="3" style="text-align: center;border-right: 2px solid black">Constante</td>
        <td colspan="3" style="text-align: center;border-right: 2px solid black">Constante</td>
        <td colspan="3" style="text-align: center;border-right: 2px solid black">Trend</td>
    </tr>
    <tr style="text-align: center;border-bottom: 2px solid black">
        <td style="text-align: center;border-right: 2px solid black">T</td>
        <td style="text-align: center;border-right: 2px dotted black">1%</td>
        <td style="text-align: center;border-right: 2px dotted black">5%</td>
        <td style="text-align: center;border-right: 2px solid black">10%</td>
        <td style="text-align: center;border-right: 2px dotted black">1%</td>
        <td style="text-align: center;border-right: 2px dotted black">5%</td>
        <td style="text-align: center;border-right: 2px solid black">10%</td>
        <td style="text-align: center;border-right: 2px dotted black">1%</td>
        <td style="text-align: center;border-right: 2px dotted black">5%</td>
        <td style="text-align: center;border-right: 2px solid black">10%</td>
    </tr>
    <tr>
        <td style="text-align: center;border-right: 2px solid black">100</td>
        <td style="text-align: center;border-right: 2px dotted black">3.22</td>
        <td style="text-align: center;border-right: 2px dotted black">2.54</td>
        <td style="text-align: center;border-right: 2px solid black">2.17</td>
        <td style="text-align: center;border-right: 2px dotted black">3.78</td>
        <td style="text-align: center;border-right: 2px dotted black">3.11</td>
        <td style="text-align: center;border-right: 2px solid black">2.73</td>
        <td style="text-align: center;border-right: 2px dotted black">3.53</td>
        <td style="text-align: center;border-right: 2px dotted black">2.79</td>
        <td style="text-align: center;border-right: 2px solid black">2.38</td>
    </tr>
    <tr>
        <td style="text-align: center;border-right: 2px solid black">250</td>
        <td style="text-align: center;border-right: 2px dotted black">3.19</td>
        <td style="text-align: center;border-right: 2px dotted black">2.53</td>
        <td style="text-align: center;border-right: 2px solid black">2.16</td>
        <td style="text-align: center;border-right: 2px dotted black">3.74</td>
        <td style="text-align: center;border-right: 2px dotted black">3.09</td>
        <td style="text-align: center;border-right: 2px solid black">2.73</td>
        <td style="text-align: center;border-right: 2px dotted black">3.49</td>
        <td style="text-align: center;border-right: 2px dotted black">2.79</td>
        <td style="text-align: center;border-right: 2px solid black">2.38</td>
    </tr>
    <tr>
        <td style="text-align: center;border-right: 2px solid black">500</td>
        <td style="text-align: center;border-right: 2px dotted black">3.18</td>
        <td style="text-align: center;border-right: 2px dotted black">2.52</td>
        <td style="text-align: center;border-right: 2px solid black">2.16</td>
        <td style="text-align: center;border-right: 2px dotted black">3.72</td>
        <td style="text-align: center;border-right: 2px dotted black">3.08</td>
        <td style="text-align: center;border-right: 2px solid black">2.72</td>
        <td style="text-align: center;border-right: 2px dotted black">3.48</td>
        <td style="text-align: center;border-right: 2px dotted black">2.78</td>
        <td style="text-align: center;border-right: 2px solid black">2.38</td>
    </tr>
    <tr>
        <td style="text-align: center;border-right: 2px solid black">$\infty$</td>
        <td style="text-align: center;border-right: 2px dotted black">3.18</td>
        <td style="text-align: center;border-right: 2px dotted black">2.52</td>
        <td style="text-align: center;border-right: 2px solid black">2.16</td>
        <td style="text-align: center;border-right: 2px dotted black">3.71</td>
        <td style="text-align: center;border-right: 2px dotted black">3.08</td>
        <td style="text-align: center;border-right: 2px solid black">2.72</td>
        <td style="text-align: center;border-right: 2px dotted black">3.46</td>
        <td style="text-align: center;border-right: 2px dotted black">2.78</td>
        <td style="text-align: center;border-right: 2px solid black">2.38</td>
    </tr>
</table>
</span>

<span style='line-height:1.5;font-size: 17px;font-family:"Times New Roman"'>
    
- Valeurs critiques du test Dickey-Fuller $\rho = 1 \Leftrightarrow \phi = 0$

<table>
    <col width="50">
    <col width="70">
    <col width="70">
    <col width="70">
    <col width="70">
    <col width="70">
    <col width="70">
    <col width="70">
    <col width="70">
    <col width="70">
    <tr>
        <td style="text-align: center;border-right: 2px solid black"></td>
        <td colspan="3" style="text-align: center;border-right: 2px solid black"><i>Modèle [1]</i></td>
        <td colspan="3" style="text-align: center;border-right: 2px solid black"><i>Modèle [2]</i></td>
        <td colspan="3" style="text-align: center;border-right: 2px solid black"><i>Modèle [3]</i></td>
    </tr>
    <tr style="text-align: center;border-bottom: 2px solid black">
        <td style="text-align: center;border-right: 2px solid black">T</td>
        <td style="text-align: center;border-right: 2px dotted black">1%</td>
        <td style="text-align: center;border-right: 2px dotted black">5%</td>
        <td style="text-align: center;border-right: 2px solid black"> 10%</td>
        <td style="text-align: center;border-right: 2px dotted black">1%</td>
        <td style="text-align: center;border-right: 2px dotted black">5%</td>
        <td style="text-align: center;border-right: 2px solid black">10%</td>
        <td style="text-align: center;border-right: 2px dotted black">1%</td>
        <td style="text-align: center;border-right: 2px dotted black">5%</td>
        <td style="text-align: center;border-right: 2px solid black">10%</td>
    </tr>
    <tr>
        <td style="text-align: center;border-right: 2px solid black">100</td>
        <td style="text-align: center;border-right: 2px dotted black">-2.60</td>
        <td style="text-align: center;border-right: 2px dotted black">-1.95</td>
        <td style="text-align: center;border-right: 2px solid black">-1.61</td>
        <td style="text-align: center;border-right: 2px dotted black">-3.51</td>
        <td style="text-align: center;border-right: 2px dotted black">-2.89</td>
        <td style="text-align: center;border-right: 2px solid black">-2.58</td>
        <td style="text-align: center;border-right: 2px dotted black">-4.04</td>
        <td style="text-align: center;border-right: 2px dotted black">-3.45</td>
        <td style="text-align: center;border-right: 2px solid black">-3.15</td>
    </tr>
    <tr>
        <td style="text-align: center;border-right: 2px solid black">250</td>
        <td style="text-align: center;border-right: 2px dotted black">-2.58</td>
        <td style="text-align: center;border-right: 2px dotted black">-1.95</td>
        <td style="text-align: center;border-right: 2px solid black">-1.62</td>
        <td style="text-align: center;border-right: 2px dotted black">-3.46</td>
        <td style="text-align: center;border-right: 2px dotted black">-2.88</td>
        <td style="text-align: center;border-right: 2px solid black">-2.57</td>
        <td style="text-align: center;border-right: 2px dotted black">-3.99</td>
        <td style="text-align: center;border-right: 2px dotted black">-3.43</td>
        <td style="text-align: center;border-right: 2px solid black">-3.13</td>
    </tr>
    <tr>
        <td style="text-align: center;border-right: 2px solid black">500</td>
        <td style="text-align: center;border-right: 2px dotted black">-2.58</td>
        <td style="text-align: center;border-right: 2px dotted black">-1.95</td>
        <td style="text-align: center;border-right: 2px solid black">-1.62</td>
        <td style="text-align: center;border-right: 2px dotted black">-3.44</td>
        <td style="text-align: center;border-right: 2px dotted black">-2.87</td>
        <td style="text-align: center;border-right: 2px solid black">-2.57</td>
        <td style="text-align: center;border-right: 2px dotted black">-3.98</td>
        <td style="text-align: center;border-right: 2px dotted black">-3.42</td>
        <td style="text-align: center;border-right: 2px solid black">-3.13</td>
    </tr>
    <tr>
        <td style="text-align: center;border-right: 2px solid black">$\infty$</td>
        <td style="text-align: center;border-right: 2px dotted black">-2.58</td>
        <td style="text-align: center;border-right: 2px dotted black">-1.95</td>
        <td style="text-align: center;border-right: 2px solid black">-1.62</td>
        <td style="text-align: center;border-right: 2px dotted black">-3.43</td>
        <td style="text-align: center;border-right: 2px dotted black">-2.86</td>
        <td style="text-align: center;border-right: 2px solid black">-2.57</td>
        <td style="text-align: center;border-right: 2px dotted black">-3.96</td>
        <td style="text-align: center;border-right: 2px dotted black">-3.41</td>
        <td style="text-align: center;border-right: 2px solid black">-3.12</td>
    </tr>
</table>    
    
    
    
    
</span>

Pour analyser la stationnarité d'une serie temporelle, il faut effectuer le test d'ADF pour chaque variable. 

### ADF - CAC40 en niveau

In [16]:
# On commence par le modèle avec tendance et constante :
"""On stock les résultats du test dans une variable ici nommée adf1:
    - On appelle la fonction de test adf
    - On spécifie la variable pour laquelle on veut faire le test. Ici : ford
    - On donne en option le modèle. Ici on commence par constante et tendance, donc : "ct"
    - Le nombre de LAG est choisi via automatiquement sous le critère d'information AIC
    - On demande à stocker les résultats pour pouvoir regarder la significativité de la tendance
"""

adf1 = adfuller(df['CAC40'], regression='ct', autolag='AIC', regresults=True)

In [17]:
adf1[3].resols.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.073
Model:                            OLS   Adj. R-squared:                  0.061
Method:                 Least Squares   F-statistic:                     6.358
Date:                Thu, 30 Nov 2023   Prob (F-statistic):            0.00220
Time:                        10:36:55   Log-Likelihood:                -1129.1
No. Observations:                 165   AIC:                             2264.
Df Residuals:                     162   BIC:                             2274.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
x1            -0.1384      0.039     -3.537      0.001      -0.216      -0.061
const        438.1519    127.612      3.433      0.001     186.155     690.148
x2             3.1516      0.922      3.417      0.001       1.330       4.973
==============================================================================
Omnibus:                       14.507   Durbin-Watson:                   1.987
Prob(Omnibus):                  0.001   Jarque-Bera (JB):               28.544
Skew:                          -0.381   Prob(JB):                     6.34e-07
Kurtosis:                       4.890   Cond. No.                     3.61e+04
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 3.61e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

- On commence par tester la variable X2 avec le modèle 3 (contante et tendance)
- La statistique de test associé à la variable X2 est de 3,417. La valeur tabulée de la tendnace du test d'ADF pour le modèle 3 est de 2,79 (pour 165 observations)
- 3,417 > 2,79 => tendance significative car t stat > valeure tabulée
- On ne rejete donc pas le modèle 3 avec tendance et constante. 

La variable est non stationnaire en tendance 

=> Pour la variable du CAC 40, la tendance est bien significative. Nous allons donc maintenant voir si nous avons une racine unitaire

In [22]:
adf1 = adfuller(df['CAC40'], regression='n', autolag='AIC', regresults=True)

In [23]:
adf1

(0.789999477104676,
 0.8832560010507973,
 {'1%': -2.5794235261707987,
  '5%': -1.942744527812561,
  '10%': -1.6153043444917494},
 <statsmodels.stats.diagnostic.ResultsStore at 0x12bc39250>)

- Nous pouvons voir que 0,789 > -1,95, 𝑡𝜑>𝑡𝐷𝐹  on ne rejette pas 𝐻0 : Racine unitaire
- Donc la série est non stationnaire, il faut la stationnariser en faisant les taux de croissance 

In [24]:
df["rCAC40"] = 100 * ( df["CAC40"]/df["CAC40"].shift(1) - 1 )
df["TxChange"] = 100 * ( df["TxChange"]/df["TxChange"].shift(1) - 1 )
df["rSocieteGenerale"] = 100 * ( df["SocieteGenerale"]/df["SocieteGenerale"].shift(1) - 1 )

df.head()

/var/folders/sg/frcwz3bs4nqf585cn0brfmjw0000gn/T/ipykernel_1012/4123414217.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["rCAC40"] = 100 * ( df["CAC40"]/df["CAC40"].shift(1) - 1 )
/var/folders/sg/frcwz3bs4nqf585cn0brfmjw0000gn/T/ipykernel_1012/4123414217.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["TxChange"] = 100 * ( df["TxChange"]/df["TxChange"].shift(1) - 1 )
/var/folders/sg/frcwz3bs4nqf585cn0brfmjw0000gn/T/ipykernel_1012/4123414217.py:3: SettingWithCopyWarning: 
A value is trying to

Date        CAC40  TxChange  SocieteGenerale    rCAC40  \
0  2010-01-01  3739.459961       NaN        20.472494       NaN   
1  2010-02-01  3708.800049 -1.722473        19.596941 -0.819902   
2  2010-03-01  3974.010010 -0.952317        22.587410  7.150829   
3  2010-04-01  3816.989990 -1.554544        19.596941 -3.951173   
4  2010-05-01  3507.560059 -7.596368        17.001795 -8.106648   

   rSocieteGenerale  
0               NaN  
1         -4.276729  
2         15.259876  
3        -13.239539  
4        -13.242609

In [27]:
df = df.dropna()
df.head()

Date        CAC40  TxChange  SocieteGenerale    rCAC40  \
1  2010-02-01  3708.800049 -1.722473        19.596941 -0.819902   
2  2010-03-01  3974.010010 -0.952317        22.587410  7.150829   
3  2010-04-01  3816.989990 -1.554544        19.596941 -3.951173   
4  2010-05-01  3507.560059 -7.596368        17.001795 -8.106648   
5  2010-06-01  3442.889893 -0.470846        16.637993 -1.843737   

   rSocieteGenerale  
1         -4.276729  
2         15.259876  
3        -13.239539  
4        -13.242609  
5         -2.139786

### ADF - CAC40 en taux de croissance

In [75]:
adf1 = adfuller(df['rCAC40'], regression='ct', autolag='AIC', regresults=True)

On recommence le test ADF

- La statistique de test associée à la variable X2 est de 0,330. La valeur tabulée de la tendnace du test d'ADF pour le modèle 3 est de 2,79 (pour 165 observations)
- T stat < T DF=> Tendance non significative, on passe donc au modèle 2


In [74]:
adf1 = adfuller(df['rCAC40'], regression='c', autolag='AIC', regresults=True)

adf1[3].resols.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.525
Model:                            OLS   Adj. R-squared:                  0.522
Method:                 Least Squares   F-statistic:                     179.2
Date:                Thu, 30 Nov 2023   Prob (F-statistic):           5.44e-28
Time:                        12:09:23   Log-Likelihood:                -489.34
No. Observations:                 164   AIC:                             982.7
Df Residuals:                     162   BIC:                             988.9
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
x1            -1.0522      0.079    -13.385      0.000      -1.207      -0.897
const          0.5192      0.378      1.374      0.171      -0.227       1.265
==============================================================================
Omnibus:                        8.966   Durbin-Watson:                   1.991
Prob(Omnibus):                  0.011   Jarque-Bera (JB):               17.939
Skew:                          -0.101   Prob(JB):                     0.000127
Kurtosis:                       4.608   Cond. No.                         4.84
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

On test maintenant la significativité de la constante:

- T stat de la constance = 1,374 et valeur tabulée de la constance dans modèle 2 = 2,54
- T stat < T DF => Constante non significative, on passe au modèle 1 sans constante ni tendance

In [73]:
adf1 = adfuller(df['rCAC40'], regression='n', autolag='AIC', regresults=True)

adf1

(-13.278014861359509,
 3.677265811037128e-24,
 {'1%': -2.579507779595479,
  '5%': -1.9427558378886698,
  '10%': -1.6152956447599425},
 <statsmodels.stats.diagnostic.ResultsStore at 0x12fdf5550>)

- T stat = -13,27 et valeur tabulée = -1,94
- T stat < T DF => On rejette H0 => notre serie est enfin stationnaire et la série CAC40 est dite intégrée d'odre 1 puisqu'il a fallu la différencier une fois pour la stationnariser.

### ADF - TxChange en niveau

In [18]:
# On commence par le modèle avec tendance et constante :
"""On stock les résultats du test dans une variable ici nommée adf1:
    - On appelle la fonction de test adf
    - On spécifie la variable pour laquelle on veut faire le test. Ici : ford
    - On donne en option le modèle. Ici on commence par constante et tendance, donc : "ct"
    - Le nombre de LAG est choisi via automatiquement sous le critère d'information AIC
    - On demande à stocker les résultats pour pouvoir regarder la significativité de la tendance
"""

adf1 = adfuller(df['TxChange'], regression='ct', autolag='AIC', regresults=True)

In [19]:
adf1[3].resols.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.136
Model:                            OLS   Adj. R-squared:                  0.090
Method:                 Least Squares   F-statistic:                     2.946
Date:                Thu, 30 Nov 2023   Prob (F-statistic):            0.00436
Time:                        10:42:53   Log-Likelihood:                 349.02
No. Observations:                 159   AIC:                            -680.0
Df Residuals:                     150   BIC:                            -652.4
Df Model:                           8                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
x1            -0.1111      0.034     -3.275      0.001      -0.178      -0.044
x2            -0.0083      0.077     -0.108      0.914      -0.161       0.144
x3             0.1518      0.077      1.984      0.049       0.001       0.303
x4             0.0964      0.074      1.297      0.196      -0.050       0.243
x5             0.0413      0.075      0.554      0.581      -0.106       0.189
x6            -0.0402      0.075     -0.539      0.591      -0.188       0.107
x7             0.2394      0.075      3.203      0.002       0.092       0.387
const          0.1495      0.046      3.248      0.001       0.059       0.240
x8            -0.0002      8e-05     -2.737      0.007      -0.000   -6.09e-05
==============================================================================
Omnibus:                        2.572   Durbin-Watson:                   1.984
Prob(Omnibus):                  0.276   Jarque-Bera (JB):                2.259
Skew:                          -0.153   Prob(JB):                        0.323
Kurtosis:                       3.497   Cond. No.                     3.88e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 3.88e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

On recommence et on tester la variable X8 avec le modèle 3 (contante et tendance)
- La statistique de test associé à la variable X8 est de -2,737. La valeur tabulée de la tendnace du test d'ADF pour le modèle 3 est de 2,79 (164 observations)
- -2,737 < 2,79 => tendance non significative car t stat < valeure tabulée 
- On rejete donc le modèle 3 avec tendance et constante et on passe donc au modèle 2 sans tendance avec constante.

On test maintenant la significativité de la constante: 

In [33]:
adf1 = adfuller(df['TxChange'], regression='c', autolag='AIC', regresults=True)
adf1[3].resols.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.528
Model:                            OLS   Adj. R-squared:                  0.525
Method:                 Least Squares   F-statistic:                     181.1
Date:                Thu, 30 Nov 2023   Prob (F-statistic):           3.39e-28
Time:                        11:03:38   Log-Likelihood:                -379.65
No. Observations:                 164   AIC:                             763.3
Df Residuals:                     162   BIC:                             769.5
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
x1            -1.0546      0.078    -13.459      0.000      -1.209      -0.900
const         -0.1298      0.193     -0.673      0.502      -0.510       0.251
==============================================================================
Omnibus:                        4.382   Durbin-Watson:                   1.997
Prob(Omnibus):                  0.112   Jarque-Bera (JB):                5.074
Skew:                          -0.155   Prob(JB):                       0.0791
Kurtosis:                       3.804   Cond. No.                         2.47
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""


- T stat de la constance = -0,673 et valeur tabulée de la constance dans modèle 2 = 2,54
- T stat < T DF => Constante non significative, on passe au modèle 1 sans constante ni tendance

In [35]:
adf1 = adfuller(df['TxChange'], regression='n', autolag='AIC', regresults=True)
adf1

(-13.464835114201934,
 1.8207113626160667e-24,
 {'1%': -2.579507779595479,
  '5%': -1.9427558378886698,
  '10%': -1.6152956447599425},
 <statsmodels.stats.diagnostic.ResultsStore at 0x12ed29350>)

- T stat = -13,46 et valeur tabulée = -1,94
- 𝑡𝜑<𝑡𝐷𝐹 on rejette 𝐻0 : la série est stationnaire

=> Pas besoin de la stationnariser 

### ADF - SocieteGenerale en niveau

In [49]:
# On commence par le modèle avec tendance et constante :
"""On stock les résultats du test dans une variable ici nommée adf1:
    - On appelle la fonction de test adf
    - On spécifie la variable pour laquelle on veut faire le test. Ici : ford
    - On donne en option le modèle. Ici on commence par constante et tendance, donc : "ct"
    - Le nombre de LAG est choisi via automatiquement sous le critère d'information AIC
    - On demande à stocker les résultats pour pouvoir regarder la significativité de la tendance
"""

ad1 = adfuller(df['SocieteGenerale'], regression='ct', autolag='AIC', regresults=True)

In [50]:
ad1[3].resols.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.046
Model:                            OLS   Adj. R-squared:                  0.034
Method:                 Least Squares   F-statistic:                     3.878
Date:                Thu, 30 Nov 2023   Prob (F-statistic):             0.0227
Time:                        11:19:53   Log-Likelihood:                -353.00
No. Observations:                 164   AIC:                             712.0
Df Residuals:                     161   BIC:                             721.3
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
x1            -0.0913      0.033     -2.783      0.006      -0.156      -0.027
const          1.5724      0.660      2.384      0.018       0.270       2.875
x2             0.0028      0.004      0.777      0.438      -0.004       0.010
==============================================================================
Omnibus:                        9.439   Durbin-Watson:                   1.879
Prob(Omnibus):                  0.009   Jarque-Bera (JB):                9.526
Skew:                          -0.512   Prob(JB):                      0.00854
Kurtosis:                       3.588   Cond. No.                         389.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

- La statistique de test associé à la variable X1 est de 0,777. La valeur tabulée de la tendnace du test d'ADF pour le modèle 3 est de 2,79 (164 observations)
- 0,777 < 2,79 => tendance non significative car t stat < valeure tabulée 
- On rejete donc le modèle 3 avec tendance et constante et on passe donc au modèle 2 sans tendance avec constante.

In [39]:
adf1 = adfuller(df['SocieteGenerale'], regression='c', autolag='AIC', regresults=True)
adf1[3].resols.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.042
Model:                            OLS   Adj. R-squared:                  0.036
Method:                 Least Squares   F-statistic:                     7.169
Date:                Thu, 30 Nov 2023   Prob (F-statistic):            0.00818
Time:                        11:11:28   Log-Likelihood:                -353.30
No. Observations:                 164   AIC:                             710.6
Df Residuals:                     162   BIC:                             716.8
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
x1            -0.0850      0.032     -2.677      0.008      -0.148      -0.022
const          1.6784      0.645      2.604      0.010       0.406       2.951
==============================================================================
Omnibus:                        8.672   Durbin-Watson:                   1.884
Prob(Omnibus):                  0.013   Jarque-Bera (JB):                8.541
Skew:                          -0.501   Prob(JB):                       0.0140
Kurtosis:                       3.494   Cond. No.                         80.0
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

- T stat de la constance = 2,604 et valeur tabulée de la constance dans modèle 2 = 2,54
- T stat > T DF => Constante significative
- On regarde si on a une racine unitaire

In [41]:
adf1 = adfuller(df['SocieteGenerale'], regression='n', autolag='AIC', regresults=True)
adf1

(-0.6143981350200205,
 0.44859909981481105,
 {'1%': -2.579507779595479,
  '5%': -1.9427558378886698,
  '10%': -1.6152956447599425},
 <statsmodels.stats.diagnostic.ResultsStore at 0x12ed4f850>)

- 𝑡𝜑 > 𝑡𝐷𝐹 on rejette 𝐻0 : la série est non stationnaire, il faut la stationnariser avec les taux de croissance

### ADF - SocieteGenerale en taux de croissance

In [51]:
# On commence par le modèle avec tendance et constante :
"""On stock les résultats du test dans une variable ici nommée adf1:
    - On appelle la fonction de test adf
    - On spécifie la variable pour laquelle on veut faire le test. Ici : ford
    - On donne en option le modèle. Ici on commence par constante et tendance, donc : "ct"
    - Le nombre de LAG est choisi via automatiquement sous le critère d'information AIC
    - On demande à stocker les résultats pour pouvoir regarder la significativité de la tendance
"""

adf1 = adfuller(df['rSocieteGenerale'], regression='ct', autolag='AIC', regresults=True)

In [52]:
adf1[3].resols.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.490
Model:                            OLS   Adj. R-squared:                  0.483
Method:                 Least Squares   F-statistic:                     77.28
Date:                Thu, 30 Nov 2023   Prob (F-statistic):           2.97e-24
Time:                        11:20:30   Log-Likelihood:                -637.66
No. Observations:                 164   AIC:                             1281.
Df Residuals:                     161   BIC:                             1291.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
x1            -0.9810      0.079    -12.431      0.000      -1.137      -0.825
const          0.4636      1.871      0.248      0.805      -3.231       4.158
x2             0.0033      0.020      0.170      0.865      -0.036       0.042
==============================================================================
Omnibus:                        7.490   Durbin-Watson:                   1.977
Prob(Omnibus):                  0.024   Jarque-Bera (JB):               13.170
Skew:                           0.097   Prob(JB):                      0.00138
Kurtosis:                       4.375   Cond. No.                         191.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

- La statistique de test associé à la variable X2 est de 0,170. La valeur tabulée de la tendnace du test d'ADF pour le modèle 3 est de 2,79 (164 observations)
- 0,170 < 2,79 => tendance non significative car t stat < valeure tabulée 
- On rejete donc le modèle 3 avec tendance et constante et on passe donc au modèle 2 sans tendance avec constante.

In [53]:
adf1 = adfuller(df['rSocieteGenerale'], regression='c', autolag='AIC', regresults=True)
adf1[3].resols.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.490
Model:                            OLS   Adj. R-squared:                  0.487
Method:                 Least Squares   F-statistic:                     155.5
Date:                Thu, 30 Nov 2023   Prob (F-statistic):           1.91e-25
Time:                        11:21:46   Log-Likelihood:                -637.68
No. Observations:                 164   AIC:                             1279.
Df Residuals:                     162   BIC:                             1286.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
x1            -0.9806      0.079    -12.469      0.000      -1.136      -0.825
const          0.7389      0.930      0.794      0.428      -1.098       2.576
==============================================================================
Omnibus:                        7.554   Durbin-Watson:                   1.978
Prob(Omnibus):                  0.023   Jarque-Bera (JB):               13.353
Skew:                           0.097   Prob(JB):                      0.00126
Kurtosis:                       4.384   Cond. No.                         11.9
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

- T stat de la constance = 0,794 et valeur tabulée de la constance dans modèle 2 = 2,54
- T stat < T DF => Constante non significative
- On passe au modèle 1 sans tendance ni constante 

In [54]:
adf1 = adfuller(df['rSocieteGenerale'], regression='n', autolag='AIC', regresults=True)
adf1

(-12.457538492571107,
 9.932477392117726e-23,
 {'1%': -2.579507779595479,
  '5%': -1.9427558378886698,
  '10%': -1.6152956447599425},
 <statsmodels.stats.diagnostic.ResultsStore at 0x12ee5f490>)

- 𝑡𝜑 (-12,45) < 𝑡𝐷𝐹 (-1,94) on rejette 𝐻0 : la série est non stationnaire

## Analyse de la multicolinéarité 

Nous allons maintenant examiner la présence de multicolinéarité dans notre modèle. La multicolinéarité est une situation dans laquelle deux ou plusieurs variables indépendantes dans un modèle de régression sont fortement corrélées entre elles. Cela peut entraîner plusieurs problèmes lors de l'ajustement d'un modèle de régression. 

In [56]:
from statsmodels.stats.outliers_influence import variance_inflation_factor as VIF

In [59]:
l_explicatives =  ['TxChange','rCAC40'] 
for n,var in enumerate(l_explicatives):
    print('le VIF de la variable', var, ' est : ',VIF(df[l_explicatives].dropna().values,n))

le VIF de la variable TxChange  est :  1.0635943443557883
le VIF de la variable rCAC40  est :  1.0635943443557885


On peut voir que pour les 2 varibales, la VIF est inférieur à 2 ce qui montre qu'il y a une multicolinéarité très faible voire inexistante 

## Estimation par les  MCO 

In [60]:
import statsmodels.api as sm
X = df[["TxChange","rCAC40"]].dropna()
X = sm.add_constant(X)
y = df["SocieteGenerale"].dropna()
modele2 = sm.OLS(y,X)
resultats2 = modele2.fit()
print(resultats2.summary())

                            OLS Regression Results                            
Dep. Variable:        SocieteGenerale   R-squared:                       0.019
Model:                            OLS   Adj. R-squared:                  0.007
Method:                 Least Squares   F-statistic:                     1.552
Date:                Thu, 30 Nov 2023   Prob (F-statistic):              0.215
Time:                        11:28:40   Log-Likelihood:                -502.93
No. Observations:                 165   AIC:                             1012.
Df Residuals:                     162   BIC:                             1021.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         19.6034      0.404     48.518      0.0

In [1]:
residus2=resultats2.resid
plt.plot(residus2)

NameError: name 'resultats2' is not defined

Commentaires :
- Les résidus semble être de moyenne constante et égale à 0
- En revanche leur variance n'est pas constante du tout 


## Analyse de l'hétéroscédasticité

- Homoscédasticité : pour tous les individus, la variance est la même
- Hétéroscédasticité : pour au moins un individu, la variance est différente 

Faut tester s’il y a hétéro ou pas : Test de Breuch Pagan : 

- S’il n’y a pas hétéro des perturbations : MCO sont BLUE. On garde donc les estimations des MCO pour les tests, IDC…
- S’il y a hétéro des perturbations : MCO sont pas BLUE et l’estimation des écarts types des coeff sont biaisés. Faut corriger le modèle avec White 

=> Tester la présence d’hétéro avant de faire les IDC, test de student, fisher… 


In [62]:
from statsmodels.stats.diagnostic import het_breuschpagan, het_white

In [63]:
bp = het_breuschpagan(residus2, X)
print(bp)

(1.2021203511795648, 0.5482301077096183, 0.5944628138917782, 0.5530582031905857)


Commentaires: 
- Avec une p-valeur de 0.55, qui est supérieure à 0.05, on ne rejete pas l'hypothèse nulle d'homoscédasticité.
- Ici, il n’y a pas hétéroscédasticité des perturbations : MCO sont BLUE. On garde donc les estimations des MCO 

## Analyse de l'autocorrélation des perturbations 

Une corrélation entre les résidus suggère qu'il y a une sorte de relation entre les erreurs et cela peut avoir des mauvaises implications sur la façon dont le modèle fonctionne et sur la fiabilité des prédictions.

Pour vérifier la présence d'autocorrélation des perturbations, on doit effectuer le test de Breuch Godfrey

In [64]:
from statsmodels.stats.diagnostic import acorr_breusch_godfrey

In [66]:
bg = acorr_breusch_godfrey(resultats2) 
print(bg)

(148.97896428675924, 6.045450854435855e-27, 141.3441863366722, 1.126651428072943e-71)


- La statistique de test de 148,97 indique une autocorrélation significative des résidus.
- La p-valeur associée à cette statistique est importante (6,045) ce qui est supérieure au seuil significatif de 0,05, ce qui suggère que l'on ne rejete pas l'hypothèse nulle d'absence d'autocorrélation des résidus.
- Il n'y a donc pas d'autocorrélation 

##### Il n'y a donc pas besoins de corriger ni l'autoccorrélation, ni l'hétéroscédasticité pour ce modèle, nous pouvons donc directement estimer notre modèle

In [67]:
import statsmodels.api as sm
X = df[["TxChange","rCAC40"]].dropna()
X = sm.add_constant(X)
y = df["SocieteGenerale"].dropna()
modele2 = sm.OLS(y,X)
resultats2 = modele2.fit()
print(resultats2.summary())

                            OLS Regression Results                            
Dep. Variable:        SocieteGenerale   R-squared:                       0.019
Model:                            OLS   Adj. R-squared:                  0.007
Method:                 Least Squares   F-statistic:                     1.552
Date:                Thu, 30 Nov 2023   Prob (F-statistic):              0.215
Time:                        11:45:57   Log-Likelihood:                -502.93
No. Observations:                 165   AIC:                             1012.
Df Residuals:                     162   BIC:                             1021.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         19.6034      0.404     48.518      0.0

##### Commentaires:



- R-squared : On voit que le R squared est très très faible (0,019) ce qui signifie que le modèle explique 1,9% de la variabilité de la variable dépendante.

Test de significativité de Student: 
- La constante (const) est statistiquement différente de zéro, car sa p-valeur est très nulle (0.000).
- La variable du Taux de change a un coefficient de 0,1339 avec une p-valeur moyenne (0.429), il est donc difficile de juger sa significativité  
- La variable du CAC40 a un coefficient de 0.1148 avec une p-valeur faible (0.187), indiquant qu'elle est statistiquement significative.

Test de significativité globale du modèle avec Fisher: 
- La statistique F de 1,552 indique que le modèle dans son ensemble n'est pas statistiquement significatif. 
- La p-valeur associée à la statistique F est de 0,215. Elle n'est pas très proche de 0 ce qui signifie que le modèle n'est pas vraiment globalement significatif  